src/VisualStudio/Core/Def/Implementation/ProjectSystem/RunningDocumentTableEventSink.cs

```diff
@@ -0,0 +1,243 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.ComponentModel.Composition;
+using Microsoft.CodeAnalysis.Editor.Shared.Utilities;
+using Microsoft.VisualStudio.Editor;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.TextManager.Interop;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem
+{
+    /// <summary>
+    /// Class to register with the RDT and forward RDT events.
+    /// Handles common conditions before sending out notifications.
+    /// </summary>
+    [Export(typeof(RunningDocumentTableEventSink))]
+    internal class RunningDocumentTableEventSink : IVsRunningDocTableEvents3, IDisposable
+    {
+        private readonly ForegroundThreadAffinitizedObject _foregroundAffinitization;
+        private readonly IVsEditorAdaptersFactoryService _editorAdaptersFactoryService;
+        private readonly IVsRunningDocumentTable4 _runningDocumentTable;
+        private uint _runningDocumentTableEventsCookie;
+
+        public event EventHandler<RunningDocumentTableEventArgs> OnCloseDocument;
+        public event EventHandler<RunningDocumentTableEventArgs> OnRefreshDocumentContext;
+        public event EventHandler<RunningDocumentTableEventArgs> OnReloadDocumentData;
+        public event EventHandler<RunningDocumentTableInitializedEventArgs> OnBeforeOpenDocument;
+        public event EventHandler<RunningDocumentTableInitializedEventArgs> OnInitializedDocument;
+        public event EventHandler<RunningDocumentTableRenamedEventArgs> OnRenameDocument;
+
+        [ImportingConstructor]
+        public RunningDocumentTableEventSink(IThreadingContext threadingContext, IVsEditorAdaptersFactoryService editorAdaptersFactoryService, SVsServiceProvider serviceProvider)
+        {
+            _foregroundAffinitization = new ForegroundThreadAffinitizedObject(threadingContext, assertIsForeground: true);
+            _runningDocumentTable = (IVsRunningDocumentTable4)serviceProvider.GetService(typeof(SVsRunningDocumentTable));
+            _editorAdaptersFactoryService = editorAdaptersFactoryService;
+
+            ((IVsRunningDocumentTable)_runningDocumentTable).AdviseRunningDocTableEvents(this, out _runningDocumentTableEventsCookie);
+        }
+
+        public int OnAfterFirstDocumentLock(uint docCookie, uint dwRDTLockType, uint dwReadLocksRemaining, uint dwEditLocksRemaining)
```

Comment:
I would make all these explicit interface implementation. I dont think we consume RunningDocumentTableEventSink supposed to ever call this directly except on who got this object for IVsRunningDocTableEvents3

Reply:
Makes sense to me, will change!

Reply:
@heejaechang I take it back - if I do it explicitly, I have to implement all methods from all three versions of the interface... is that better than just leaving it implicit?